### ISSUE TITLE
Java installation completes but shows "operation not permitted" error message

### ISSUE THREAD

**Issue by eppsilon** [caskroom/homebrew-cask#13941]
On OS X 10.11, installing Java results in an error message at the end of the output (2nd to last line below). I think this is because creating a symlink in /System runs afoul of El Capitan's new [System Integrity Protection](https://developer.apple.com/library/prerelease/ios/documentation/Security/Conceptual/System_Integrity_Protection_Guide/FileSystemProtections/FileSystemProtections.html#//apple_ref/doc/uid/TP40016462-CH2-SW1) feature.

```
brew cask install java --force
==> Caveats
[omitted]
==> Downloading http://download.oracle.com/otn-pub/java/jdk/8u60-b27/jdk-8u60-macosx-x64.dmg
Already downloaded: /Library/Caches/Homebrew/java-1.8.0_60-b27.dmg
==> Running installer for java; your password may be necessary.
==> Package installers may write to any location; options such as --appdir are ignored.
==> installer: Package name is JDK 8 Update 60
==> installer: Upgrading at base path /
==> installer: The upgrade was successful.
ln: /System/Library/Frameworks/JavaVM.framework/Versions/CurrentJDK: Operation not permitted
java staged at '/opt/homebrew-cask/Caskroom/java/1.8.0_60-b27' (2 files, 227M)
```

**Comment by bdhess** [comment]
The offending lines in the Java casks are:

```
postflight do
   ...snip...
    system '/usr/bin/sudo', '-E', '--',
      '/bin/rm', '-rf', '--', '/System/Library/Frameworks/JavaVM.framework/Versions/CurrentJDK'
   ...snip...
end
```

and

```
uninstall
   ...snip...
  :delete => [
                       ...snip...
                        '/System/Library/Frameworks/JavaVM.framework/Versions/CurrentJDK',
                        ...snip...
                       ]
   ...snip...
```

It looks like these lines were added to emulate the behavior of the legacy Apple JDK installer, in order to work around an issue that would cause the user to receive a prompt to install Java even when Oracle JDK 7 is installed.  For reference, here's the [commit](https://github.com/caskroom/homebrew-cask/commit/a60362e864c011a31a4347613b117ca4a433e422) that added them.

On a clean Yosemite install, I can't reproduce these conditions, even with early versions of JDK 7.  Installing 7u19 and then typing `java -version` at the command prompt happily spits out the expected java version output, rather than popping up the prompt to install.  I can only assume that Apple built a workaround for this issue in Yosemite (or earlier), that causes their Java wrappers to search for Java versions in /Library, rather than expecting the symlink in /System/Library.

For fuller details on how this interacts with JDK 6:  Installing the Apple JDK 6 results in a new symlink being created in the JavaVM.framework to point to the newly installed Java 6.  Installing Oracle JDK 7 or 8 does not result in a change to the symlink in /System.  I.e., graphically installing Apple JDK 6 and then subsequently installing Oracle JDK 7 and 8 using the .pkg installers still results in the framework symlink pointing to JDK 6.  This doesn't seem to have much practical effect on the system, at least on Yosemite.  The Apple-provided wrappers at /usr/bin/java, javac, etc, seem to prefer to use the latest installed version by default, but will use another version if specified in the JAVA_HOME environment variable.

In short-- I think the appropriate change would be to cause these blocks to execute only if running on a pre-Yosemite OS.  If there's agreement, I'd be happy to submit a patch.

**Comment by vitorgalvao** [comment]
@bdhess Thank you for doing the research. Sounds good to me, you can go ahead with the patch.

**Comment by Miladiir** [comment]
Is there any kind of work around available? I really don't want to install an ancient version of java.

**Comment by nickroberts** [comment]
I'm having issues with this, a well. On a new MBP, with an upgrade to EL Cap, trying o now install java...I get errors when trying to link the package.

```
==> installer: Package name is JDK 8 Update 60
==> installer: Upgrading at base path /
==> installer: The upgrade was successful.
ln: /System/Library/Frameworks/JavaVM.framework/Versions/CurrentJDK: Operation not permitted
🍺  java staged at '/opt/homebrew-cask/Caskroom/java/1.8.0_60-b27' (2 files, 227M)
```

Any ideas on a fix for this?

**Comment by vitorgalvao** [comment]
@nickroberts. Please read the issue before repeating stuff. Your exact issue is here, as is a proposed solution.

Wating for the mentioned patch from @bdhess or someone else that wants to test it and submit it.

**Comment by nickroberts** [comment]
I did read it. I tried to edit the local files and remove what the issue was, and the issue was still present. I'm not sure if I did it exactly right, or not. Is there documentation on how to point a cask to your local machine, so I can create a fork and create the patch? Thanks!

**Comment by vitorgalvao** [comment]
[CONTRIBUTING.md](https://github.com/caskroom/homebrew-cask/blob/master/CONTRIBUTING.md).

**Comment by bdhess** [comment]
Please pardon the first couple of erroneous commits there.  Opened PR #14233 with this fix.  If accepted, I can make a similar change to the `java7` cask in `homebrew-versions`.

**Comment by FractalizeR** [comment]
What can be done until this patch is merged?

**Comment by vitorgalvao** [comment]
Fixed by @bdhess.
